### PR TITLE
[Bifrost] Bifrost appends will wait for reconfiguration

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] Bifrost appends will wait for reconfiguration

Bifrost appends will wait for reconfiguration to complete. Additionally, log metadata will now handle empty segments correctly. An empty segment will be removed from the chain and replaced with the newly created one but with a higher segment index.

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1772).
* #1776
* #1775
* __->__ #1772

### CODE REVIEW

File: crates/bifrost/src/bifrost.rs [hunk 1]
```diff
@@ -347,6 +417,32 @@ impl BifrostInner {
...
+    async fn wait_next_unsealed_loglet(
+        &self,
+        log_id: LogId,
+        sealed_segment: SegmentIndex,
+        retry_iter: &mut RetryIter,
+    ) -> Result<LogletWrapper> {
+        let start = Instant::now();
+        for sleep_dur in retry_iter.by_ref() {
+            self.fail_if_shutting_down()?;
+            tokio::time::sleep(sleep_dur).await;
```
Comment (tillrohrmann) [inline crates/bifrost/src/bifrost.rs:429]: Could we also wait for a `logs` metadata update to occur?

File: crates/types/src/config/bifrost.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ pub struct BifrostOptions {
     /// Configuration of replicated loglet provider
     pub replicated_loglet: ReplicatedLogletOptions,
 
-    /// # Retry policy
+    /// # Read Retry policy
```
Comment (tillrohrmann) [inline crates/types/src/config/bifrost.rs:47]:
```suggestion
    /// # Read retry policy
```

File: crates/types/src/config/bifrost.rs [hunk 3]
```diff
@@ -56,6 +56,15 @@ pub struct BifrostOptions {
     #[serde_as(as = "serde_with::DisplayFromStr")]
     #[cfg_attr(feature = "schemars", schemars(with = "String"))]
     pub seal_retry_interval: humantime::Duration,
+
+    /// # Append Retry policy
```
Comment (tillrohrmann) [inline crates/types/src/config/bifrost.rs:60]:
```suggestion
    /// # Append retry policy
```

File: crates/types/src/config/bifrost.rs [hunk 4]
```diff
@@ -56,6 +56,15 @@ pub struct BifrostOptions {
...
+    ///
+    /// Retry policy to use when bifrost waits for reconfiguration to complete during
+    /// append operations.
+    ///
+    /// Note that appends will be retried forever and setting `max-attempts` have no impact
+    /// on this particular retry policy.
```
Comment (tillrohrmann) [inline crates/types/src/config/bifrost.rs:66]: Don't we stop retrying in `wait_next_unsealed_loglet` and fail with `Error::LogSealed(log_id)` if the retry policy is exceeded?

**Review by tillrohrmann — APPROVED** [review]
Really good work @AhmedSoliman :-) LGTM. +1 for merging.

File: crates/types/src/config/bifrost.rs (hunk 4 above)
Comment (AhmedSoliman) [reply crates/types/src/config/bifrost.rs]: Fixed.

**Pull request merged**